update GCP blog post (#329)
kestra-io/kestra.io: content/blogs/2022-11-19-create-data-pipeline-bigquery-google-cloud.md

image: /blogs/2022-11-19-create-data-pipeline-bigquery-google-cloud.png
---

This blog post explores the integration of Kestra with Google Cloud Platform (GCP), specifically focusing on enhancing BigQuery operations. We will elaborate on how to automate repetitive processes, streamline data pipelines, and create resilient workflows within the GCP environment using Kestra's orchestration capabilities.

## Kestra & Big Query ##

> Google Cloud Platform (GCP) is a widely adopted cloud solution for building end-to-end data pipelines, from data warehousing to building and hosting scalable applications. BigQuery, is a fully managed and serverless data warehouse, and it’s typically used for storing and analyzing large, structured data. It acts as the main tool for several use cases, such as:

Reviewer: ```suggestion
Google Cloud Platform (GCP) is one of the major cloud providers. BigQuery is GCP's serverless data warehouse. It's typically leveraged for storing and analyzing large datasets, including use cases such as:
```

*   **Data Warehousing** with serverless query operations for data analysis.

> *   **Data transfer services** to migrate data tables to/from BigQuery from another cloud platform.

Reviewer: ```suggestion
*   **Data integration** - Data Transfer Services integration allows you to ingest data from various services, including those outside of GCP's offerings, into BigQuery.
```

> *   **Multiple language-specific** **SDKs** and **gcloud CLI** are available to interact with BigQuery.

Reviewer: ```suggestion
*   **Cross-language analytics** - GCP offers SDKs and `gcloud` CLI tool to analyze data stored in BigQuery using various programming languages.
```

> *   **Machine learning model training** and prediction at regular intervals using scheduled BigQuery ML queries.

Reviewer: ```suggestion
*   **Machine learning** - model training and prediction at regular intervals using queries specialized for ML.
```

Despite the utility of tools like the gcloud CLI and Scheduled Query for simple data processing tasks, they fall short when it comes to connecting multiple data sources and destinations, creating data modelization with dependencies, and visualizing each task.

> Kestra provides a suite of plugins for various GCP services such as Google Cloud Storage (GCS), BigQuery, VertexAI, and more. The plugins for BigQuery, in particular, facilitate the creation of ETL/ELT pipelines to various other services readily available in Kestra.

Reviewer: ```suggestion
Kestra provides a suite of plugins for various GCP services, such as Google Cloud Storage (GCS), BigQuery, VertexAI, and more. The plugins for BigQuery, in particular, make it easy to connect BigQuery data to external services beyond the boundaries of GCP.
```

### Integration of Kestra and BigQuery ###

> Kestra enables standard operations in BigQuery, such as creating and deleting datasets and tables, running queries, and importing/exporting tables to/from BigQuery and GCS, all through its BigQuery plugins.

Reviewer: ```suggestion
Kestra enables standard operations in BigQuery, such as creating and deleting datasets and tables, running queries, and importing/exporting tables to/from BigQuery and GCS, and many more, using simple plugins.
```

> ### Data Modelization

Reviewer: ```suggestion
### Data Modeling
```

Data Modeling is a process that involves creating a visual representation of data flow between different data structures. This is achieved through a series of transformation or aggregation queries on the raw dataset, with the final data used for visualization, analysis, or machine learning. Kestra's [**Query**](../plugins/plugin-gcp/tasks/bigquery/io.kestra.plugin.gcp.bigquery.Query.md) plugin for BigQuery can be used for this purpose.

For instance, you can apply complex aggregation on daily sales report data and use those data points in subsequent phases of transformations for gathering the daily sales trends.

> Here's an illustrative flow that implements data modelization on daily sales data and calculates average sales for each product category. This flow is triggered daily and executes several BigQuery queries to transform and aggregate raw sales data: 

Reviewer: ```suggestion
Here's an illustrative flow that implements data modeling on daily sales data and calculates average sales for each product category. This flow is triggered daily and executes several BigQuery queries to transform and aggregate raw sales data: 
```

      FROM `your-project.your-dataset.sales_raw`
      GROUP BY date, product_category

  - id: calculate-average-sales
    type: io.kestra.plugin.gcp.bigquery.Query
    projectId: your-gcp-project-id
    destinationTable: your-project.your-dataset.sales_avg
    writeDisposition: WRITE_TRUNCATE
    sql: |
      SELECT date, product_category, AVG(sales_count) as avg_sales
      FROM `your-project.your-dataset.sales_avg`
      GROUP BY date, product_category
     
  - id: query_latest_date
    type: io.kestra.plugin.gcp.bigquery.Query
    projectId: your-gcp-project-id
    fetchOne: true
    sql: |
      SELECT MAX(added_date) AS date
      FROM `kestra-dev.ETL_demo.raw_data`

  - id: query_new_data
    type: io.kestra.plugin.gcp.bigquery.Query
    projectId: your-gcp-project-id
    destinationTable: your-project.your-dataset.sales_avg
    writeDisposition: WRITE_APPEND
    sql: |
      SELECT *
      FROM `kestra-dev.ETL_demo.source`
      WHERE added_date > date('{{ outputs.query_latest_date.row.date }}')

triggers:
  - id: schedule
    type: io.kestra.core.models.triggers.types.Schedule
    cron: "0 0 * * *"

```

> Here we have four tasks that each run a BigQuery query.

Reviewer: ```suggestion
Here we have four tasks, each executing a query within a BigQuery cloud data warehouse.
```

printHeader: true
```

> Inputs for flow execution in Kestra can be provided either via the UI or Curl. A complete API with a sample Curl command for automating the execution from another application is shown below:

Reviewer: ```suggestion
Inputs for flow execution in Kestra can be provided either via the UI or when executing the flow via an API call, for example using the `curl` command. Here is an example `curl` command that you can use to execute the flow from other applications:
```

    -F destinationFile="analysis_data_18"
```

### Storage Write ###

> The [StorageWrite](../plugins/plugin-gcp/tasks/bigquery/io.kestra.plugin.gcp.bigquery.StorageWrite.md) plugin, is used for importing data from diverse sources like databases, NoSQL, queues, or other plugins into BigQuery. This is especially useful when data is stored externally. Another advantage of using this plugin is to avoid quotas limitation to ingest the data while data streaming in real-time or in batch job writing.

Reviewer: ```suggestion
The [StorageWrite](../plugins/plugin-gcp/tasks/bigquery/io.kestra.plugin.gcp.bigquery.StorageWrite.md) plugin is used for importing data from various sources, including databases, distributed message queues, or other plugins, into BigQuery. This is especially useful when data is stored externally. Another advantage of using this plugin is to avoid quotas limitation during streaming and batch data ingestion.
```

    destinationTable: "kestra-dev.ETL_demo.transactions_logs"
```

> ## Event Driven Orchestration ##

Reviewer: ```suggestion
## Event-Driven Orchestration ##
```

> Kestra also allows you to create dynamic data pipelines that can be triggered based on certain events. For example, your data pipeline might need to start when a new table is ingested into BigQuery, or when a file is uploaded into a Google Cloud Storage (GCS) bucket. Kestra supports these event-driven workflows through its trigger plugins for BigQuery and GCS.

Reviewer: ```suggestion
Kestra also allows you to create dynamic data pipelines that can be triggered based on external events. For example, your data pipeline might need to start when a new table is ingested into BigQuery, or when a file is uploaded into a Google Cloud Storage (GCS) bucket. Kestra supports such event-driven workflows via trigger plugins for BigQuery and GCS.
```

With Kestra, the process of analyzing your flow (DAG) executions becomes intuitive and informative. Accessible through the "Execution" section, each execution instance is logged with specific details, such as its trigger source (manual or automated), as well as a unique Execution ID. This meticulously detailed record of each flow allows for comprehensive data lineage views encompassing Gantt charts, logs, topology, and outputs.

A key aspect of data lineage visualization is the Gantt chart, which provides a clear overview of the time taken for each task's data processing within the flow. In parallel, the Topology feature enables the user to identify and debug the sequence of tasks along with their dependencies.

![Gantt](/blogs/2022-11-19-create-data-pipeline-bigquery-google-cloud/Gantt.png)

Accessing specific information about the flow is simple, requiring only the unique execution ID. This enables efficient data lineage tracking of the flow. You can leverage Kestra's null variable to add any execution identifier to separate different executions in the flow. Furthermore, you can inject a null variable into SQL queries for enhanced data tracking.

## Next Steps ##

> Stay tuned for our upcoming posts as we delve deeper into GCP integrations, providing more granular, real-world use cases and a step-by-step guide to leverage these powerful tools.

Reviewer: ```suggestion
If you would like to see more posts diving into GCP integrations or integrations with other cloud providers, let us know. We'd love to hear your feedback and the use cases you would like us to cover next. 
```

better not to promise the next posts unless you've already written that and plan to schedule -- this phrasing seems more flexible and invites readers to be part of the process without committing to anything yet
